kernel/USER_GUIDE.md

```diff
@@ -0,0 +1,604 @@
+* Metadata columns: These are special columns that must be populated using metadata about the Parquet file. To understand how to populate such a column, first match the column name against the set of standard metadata column name constants. For example, 
+    * ``StructField`.`ROW_INDEX_COLUMN_NAME``, then you have to a generate column vector populated with the actual index of each row in the Parquet file (that is, not indexed by the possibly subset of rows return after Parquet data skipping). See the docs for more details.
+
+**Requirements and guarantees: **Any implementation must adhere to the following guarantees.
+
+
+
+* The schema of the returned ColumnarBatches must match the physical schema. 
+    * If a data column is not found and the ``StructField.isNullable = true``, then return a ColumnVector of nulls. Throw error if it is not nullable.
+* The output iterator must maintain ordering as the input iterator. That is, if file1 is before file2 in the input iterator, then columnar batches of file1 must be before those of file2 in the output iterator.
+
+**Performance suggestions:**
+
+
+
+* The representation of data as `ColumnVectors` and `ColumnarBatches` can have a significant impact on the query performance and it's best to read the Parquet file data directly into vectors and batches of the engine-native format to avoid potentially costly in-memory data format conversion. This is further discussed later.
+
+
+#### Implement ExpressionHandler
+
+The `[ExpressionHandler](https://vkorukanti.github.io/delta-kernel2/index.html?io/delta/kernel/types/StructType.html)` interface has all the methods needed for handling expressions that may be applied on columnar data. 
+
+
+##### Method getEvaluator() 
+
+[This method](https://vkorukanti.github.io/delta-kernel2/io/delta/kernel/client/ExpressionHandler.html#getEvaluator-io.delta.kernel.types.StructType-io.delta.kernel.expressions.Expression-) generates a function (called ExpressionEvaluator) that can evaluate an expression on a batch of row data to produce a result of a single column vector. To generate this function, the ``getEvaluator`()` method takes as input the expression and the schema of the ``ColumnarBatches`` of data on which the expressions will be applied.
+
+**Requirements and guarantees: **Any implementation must adhere to the following guarantees.
+
+
+
+* Implementation must handle all possible variations of expressions. If they implementation encounters an expression type that it does not know how to handle, then it must throw a standard exception.
+    * Java: [NotSupportedException](https://docs.oracle.com/javaee/7/api/javax/resource/NotSupportedException.html) 
+* The ColumnarBatches on which the generated ExpressionEvaluator is going to be used are guaranteed to have the schema provided during generation. Hence, it is safe to bind the expression evaluation logic to column ordinals instead of column names, thus making the actual evaluation faster.
```

Comment:
```suggestion
* The `ColumnarBatch`es on which the generated `ExpressionEvaluator` is going to be used are guaranteed to have the schema provided during generation. Hence, it is safe to bind the expression evaluation logic to column ordinals instead of column names, thus making the actual evaluation faster.
```